plugins/out_s3/s3.c

```diff
@@ -252,6 +253,143 @@ static flb_sds_t concat_path(char *p1, char *p2)
     return dir;
 }
 
+/* Reads in index value from metadata file and sets seq_index to value */
+static int read_seq_index(char *seq_index_file, uint64_t *seq_index)
+{
+    FILE *fp;
+    int ret;
+
+    fp = fopen(seq_index_file, "r");
+    if (fp == NULL) {
+        return -1;
+    }
+
+    ret = fscanf(fp, "%"PRIu64, seq_index);
+    if (ret != 1) {
+        return -1;
+    }
+
+    fclose(fp);
+    return 0;
+}
+
+/* Writes index value to metadata file */
+static int write_seq_index(char *seq_index_file, uint64_t seq_index)
+{
+    FILE *fp;
+    int ret;
+
+    fp = fopen(seq_index_file, "w+");
+    if (fp == NULL) {
+        return -1;
+    }
+
+    ret = fprintf(fp, "%"PRIu64, seq_index);
+    if (ret < 0) {
+        return -1;
+    }
+
+    fclose(fp);
+    return 0;
+}
+
+static int init_seq_index(void *context) {
+    int ret;
+    int len;
+    const char *tmp;
+    char *tmp_buf;
+    struct flb_s3 *ctx = context;
+
+    ctx->key_fmt_has_seq_index = FLB_TRUE;
+
+    /* Initialize local storage to store metadata files */
+    ret = s3_store_init(ctx);
+    if (ret == -1) {
+        flb_plg_error(ctx->ins, "Failed to initialize S3 storage: %s",
+                      ctx->store_dir);
+        return -1;
+    }
+
+    ctx->stream_metadata = flb_fstore_stream_create(ctx->fs, "metadata");
+    if (!ctx->stream_metadata) {
+        flb_plg_error(ctx->ins, "could not initialize metadata stream");
+        flb_fstore_destroy(ctx->fs);
+        ctx->fs = NULL;
+        return -1;
+    }
+
+    /* Construct directories and file path names */
+    tmp = ctx->stream_metadata->path;
+    ctx->metadata_dir = flb_sds_create(tmp);
+    if (ctx->metadata_dir == NULL) {
+        flb_plg_error(ctx->ins, "Failed to create metadata path");
+        flb_errno();
+        return -1;
+    }
+    tmp = "/index_metadata";
+    ctx->metadata_dir = flb_sds_cat(ctx->metadata_dir, tmp, strlen(tmp));
```

Comment:
If you look for `flb_sds_cat` in the code there is an existing pattern where its return value is set to a temporary pointer. Then you check that pointer and then set it to the final pointer. 

This is because under the hood it uses realloc, and on realloc failure, you need to free the original memory: https://stackoverflow.com/questions/1607004/does-realloc-free-the-former-buffer-if-it-fails